### ISSUE TITLE
RFC: New feature: subdivide builders into projects

### ISSUE THREAD

**Issue by p12tic** [buildbot/buildbot#4373]
I work with a buildbot instance with several unrelated "projects" with a lot of builders. Due to the number of builders navigating buildbot is a little complicated. I propose we add a projects feature:

 - there's a list of projects configured in the buildbot config
 - each builder is assigned a project
 - in the www UI we add a separate menu item where it's possible to select an active project
      - let's say just below the Home menu there's `Projects` which expands to the list of projects just like "Builds" expands to a list of menu items.
 - in the www UI we show only builders and builds that are related to the active project
 - the web UI shows the name of the active project somewhere

A very similar thing has been implemented in Unity Technologies fork of buildbot called [Katana](https://github.com/Unity-Technologies/katana) from which I gathered inspiration for this feature.

At this point I think this feature will not need any complex refactors, only a significant amount of small changes all around the codebase and especially in the frontend.

What do you think? I'm eager to implement this :-)

**Comment by ari** [comment]
buildbot already has tags which probably do what you want

**Comment by p12tic** [comment]
Tags don't work well when there are a lot of projects each of them with a lot of builders. In the buildbot instance I manage there are several hundreds of builders, the builder page is already a bit inconvenient to use. Buildbot instance at my day job handles thousands of builders, so in cases like this some way to reduce noise is essential for usability.

**Comment by p12tic** [comment]
Tags don't work well when there are a lot of projects each of them with a lot of builders. In the buildbot instance I manage there are several hundreds of builders, the builder page is already a bit inconvenient to use. Buildbot instance at my day job handles thousands of builders, so in cases like this some way to reduce noise is essential for usability.

**Comment by ari** [comment]
So you don't need a new concept of projects, just some new UI choices to display tags. Or your own custom dashboard built in Flask to do what you need.

**Comment by p12tic** [comment]
I'd say tags are unrelated to projects. What I would like to do is to have completely separate UI for each project which would not show builders, builds and tags related to different projects. There's at least one existing implementation of this feature in a fork of buildbot used by a large company so it's not like I'm suggesting something that would only be useful for me. Ideally I would just adapt their approach to the current buildbot codebase as it's proved to work.

One thing that can only be easily expressed by project subdivision is project-specific codebases. A project gets certain codebases configured and all builders within that project get them automatically.

**Comment by p12tic** [comment]
Later in the future I'd like to add UI that would limit builds shown on a certain project to certain branches in the codebases configured for that project. This way a developer can easily figure out the build status on his own branch without going through some external UI such as PR CI status on github or similar.

**Comment by tardyp** [comment]
I agree with @ari that tags are generic enough to support the use cases that you are describing. What we don't have is a UI powerful enough to make use of it.

I agree on the need for a project specific landing page
 
UI specific to a project would mean that you could specify a tag filter and it would follow your browsing history through all the buildbot pages until you switch the tag filtering.

In my understanding of buildbot, it would not really simplify things to add another project identifier

**Comment by p12tic** [comment]
@tardyp, @ari: After some thought I agree that just some UI addition to current implementation of tags is enough to solve the use case that I've listed in the proposal. Unfortunately trying to reduce its scope I did not talk about codebases at all, which is the actual reason for having the project identifier at all.

Basically, the reason for having project identifier is that it would allow us to have a UI where you could see all codebases related to project, select the branches you want and get the status of the last build of each project-specific builder on this specific set of branches. This is a workflow we use at Unity everyday and I really miss it when using the bare buildbot for my own projects. When adding the above information, does my proposal make sense?

**Comment by p12tic** [comment]
The workflow I'm describing is the following:

Have a project `P` that consists of 3 codebases, let's say `CA`, `CB` and `CC`. You have a bunch of builders, let's say `B1`, `B2`, `B3` which all need `CA`, `CB` and `CC` to build. Let's say each of the codebases have `master` branch where PRs are merged to.

I want to make some change on codebase CA. I branch off to `mybranch` branch, do some coding, push. Now I want to run the builders and see whether my changes broke anything.

I go to buildbot page for project `P`. It lists three drop downs to select revisions/branches for each of the configured codebases: `CA`, `CB` and `CC`. I select `mybranch` for `CA` and `master` for `CB` and `CC`. Buildbot shows me the list of builders configured for the project - `B1`, `B2`, `B3`. Each of them indicates that no builds have ever been ran on them. I start the builds and come back later to check out results. Unless someone started more builds for the same branches I specified, I will be sure that the builder list displays build statuses of the builds I started regardless of how many builds have been completed on other branches in the meantime.

I believe that it's possible to implement the above without cluttering the views too much for users that don't need this kind of setup. Ideally my changes wouldn't change the buildbot website at all if a user has not setup projects.

**Comment by tardyp** [comment]
I think this is not easy to add. 
Lets see what you can come up with.

**Comment by p12tic** [comment]
>  Lets see what you can come up with.

@tardyp: Just to confirm, do I correctly understand that you think that the feature makes sense in principle, just that it may turn out to be more complex than it's worth?

**Comment by tardyp** [comment]
It makes sense to have a project table which would associate with builder and codebase.

codebase is imho a bit more complex than it should.

For example the base example sounds like very mozilla or hg specific. 
```
In these days of distributed version control systems, the same codebase may appear in multiple repositories. For example, https://github.com/mozilla/mozilla-central and http://hg.mozilla.org/mozilla-release both contain the Firefox codebase, although not exactly the same code.
```
While it does make sense, I am not sure exactely if it is needed, and how this actually help resolve some releng problem.
The way it is integrated into change filters makes it quite too difficult in my opinion to craft a complex codebase based buildbot config.

I would be glad to see some more reflexion on this topic, and to improve the notion of codebase in buildbot and its integration. Having new brain look at it with a fresh mind would be a great help, I think.

Other use case of a project concept would be for authorization. Would be cool to have elevated rights to be specific to projects (rather than builders)

Some plantuml diagram might probably help to understand the relationship of project with other buildbot concepts, and convince us further that this is not something that everybody needs to hack into tags.

**Comment by p12tic** [comment]
> It makes sense to have a project table which would associate with builder and codebase.

I've looked into how Katana solves this problem I want buildbot to solve and it did not create any new tables for the project relationships. It actually turned out that buildbot already does have the concept of projects and the change sources side of things already supports it. So far it looks like that only the configuration, frontend and data layer bits are missing in the official buildbot.

> For example the base example sounds like very mozilla or hg specific.

At Unity Technologies we have a similar setup with both hg and git repositories of certain projects. The release branches live in a separate, smaller repository which is read-only for most people. There are various reasons for such split: from performance to source code sharing with external colaborators.

> Other use case of a project concept would be for authorization. Would be cool to have elevated rights to be specific to projects (rather than builders)

Agreed.

> Some plantuml diagram might probably help to understand the relationship of project with other buildbot concepts, and convince us further that this is not something that everybody needs to hack into tags.

I will see what I can do. Maybe we can later add something to the official buildbot docs.

**Comment by p12tic** [comment]
The following is a UML-like diagram of how project currently relates to other buildbot concepts up until builder.

![test](https://user-images.githubusercontent.com/1056711/47967317-a5559180-e064-11e8-838c-20fb5dd1409b.png)

Projects are currently a half-baked feature. It's not used much anywhere and for example, schedulers don't support it fully (empty project is used in various places).

I suppose for full project support we could first add support it on the builder side and then work from there, fixing scheduler support bit by bit. This would not impact regular users, because the new code paths would only be enabled if they set the new 'projects' key in the configuration. @tardyp What do you think?

**Comment by p12tic** [comment]
This has been addressed by experimental project support in Buildbot master. This will be released as Buildbot 3.9.0
